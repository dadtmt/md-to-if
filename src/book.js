import * as R from 'ramda'
import { snakeCase } from 'change-case'

export const getSceneName = R.pipe(
  R.propOr([], 'content'),
  R.head,
  R.propOr('unnamed', 'content'),
  snakeCase
)

export const splitByScene = R.pipe(
  R.converge(R.prepend, [
    R.head,
    R.pipe(
      R.tail,
      R.splitWhen(
        R.where({
          type: R.equals('heading'),
          level: R.equals(2),
        })
      )
    ),
  ]),
  R.zipObj(['heading', 'content', 'sourceLeft']),
  ({ heading, content, sourceLeft }) => ({
    scene: {
      name: getSceneName(heading),
      content: [heading, ...content],
    },
    sourceLeft,
  })
)

const book = (source, scenes = []) => {
  const { scene, sourceLeft } = splitByScene(source)
  return R.isEmpty(sourceLeft)
    ? [...scenes, scene]
    : book(sourceLeft, [...scenes, scene])
}

export default book
