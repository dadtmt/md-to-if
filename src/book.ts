import * as R from 'ramda'
import { snakeCase } from 'change-case'

// Content -> String
export const getSceneName = R.pipe(
  R.propOr([], 'content'),
  R.head,
  R.propOr('unnamed', 'content'),
  snakeCase
)

// int -> [Content] -> [[Content],[Content]]
const splitByHeading = level =>
  R.splitWhen(
    R.where({
      type: R.equals('heading'),
      level: R.equals(level),
    })
  )

const splitActions = (level, contentList, actionList = []) => {
  const [headOfContent, ...tailOfContent] = contentList
  const [actionContent, restOfContent] = splitByHeading(level)(tailOfContent)
  const { content, actions } =
    level <= 6
      ? splitContentAndActions(level + 1)(actionContent)
      : {
          content: [],
          actions: [],
        }
  return !R.isEmpty(restOfContent)
    ? splitActions(level, restOfContent, [
        ...actionList,
        {
          name: getSceneName(headOfContent),
          sceneContent: [headOfContent, ...content],
          actions,
        },
      ])
    : actionList
}

// [Content] -> { content: [Content], actions: [Scene] }
export const splitContentAndActions = (level: number) => contentAndActions => {
  const [content, actions] = splitByHeading(level)(contentAndActions)
  return { content, actions: splitActions(level, actions) }
}

// [Content] -> { scene: Scene, content: [Content] }
export const splitByScene = (level: number): Function =>
  R.pipe(
    R.converge(R.prepend, [R.head, R.pipe(R.tail, splitByHeading(level))]),
    R.zipObj(['heading', 'content', 'sourceLeft']),
    ({ content, ...rest }) => ({
      content,
      ...rest,
    }),
    ({ heading, content, sourceLeft }) => ({
      scene: {
        name: getSceneName(heading),
        sceneContent: [heading, ...content],
      },
      sourceLeft,
    })
  )

// [Content], [Scene] -> [Scene]
const book = (source, scenes = []) => {
  const { scene, sourceLeft } = splitByScene(2)(source)
  return R.isEmpty(sourceLeft)
    ? [...scenes, scene]
    : book(sourceLeft, [...scenes, scene])
}

export default book
