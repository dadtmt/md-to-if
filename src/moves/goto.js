import * as R from 'ramda'
import { incPlayedSceneCount } from './stateHelpers/playedSceneCount'

// [PlayedScene] -> State
const getLastPlayedSceneState = R.pipe(R.last, R.prop('state'))

//  Move -> String
export const getTargetSceneName = R.pipe(R.prop('target'), R.tail)

// Move -> Scene -> Boolean
export const matchTarget = move => R.propEq('name', getTargetSceneName(move))

// Move -> [Scene] -> Scene
export const getTargetedScene = move => R.find(matchTarget(move))

// String -> State -> State
const updateState = name =>
  R.pipe(R.assoc('currentSceneName', name), incPlayedSceneCount(name))

// [Scene], [PlayedScene]-> [Move -> Boolean, Move -> MovedScene]
const goto = (scenes, playedScenes) => [
  R.propEq('type', 'anchor'),
  move => {
    const lastState = getLastPlayedSceneState(playedScenes)
    const scene = getTargetedScene(move)(scenes)
    const { name } = scene

    return {
      ...scene,
      state: updateState(name)(lastState),
    }
  },
]

export default goto
