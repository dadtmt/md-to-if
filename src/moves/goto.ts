import * as R from 'ramda'
import { incPlayedSceneCount } from '../stateHelpers/playedSceneCount'
import { PlayedScene, Move } from '../player'
import { MovedScene, State } from '.'
import { BookScene } from '..'

const getState: (playedScene: PlayedScene) => State = R.prop('state')

const getLastPlayedSceneState: (playedScenes: PlayedScene[]) => State = R.pipe(
  R.last,
  getState
)

const notFoundScene: (move: Move) => BookScene = (move) => {
  return {
    actions: [],
    name: 'unknown',
    sceneContent: [
      {
        type: 'heading',
        level: 2,
        content: [
          {
            type: 'text',
            content: 'unknown'
          }
        ]
      },
      {
        type: 'paragraph',
        content: [
          {
            type: 'text',
            content: `The scene with path: /${getPath(move).join(
              '/'
            )} does not exist`
          }
        ]
      }
    ]
  }
}

const getSceneByPath =
  (move: Move) =>
  (scenes: BookScene[], path: string[]): BookScene => {
    const [firstSceneName, ...restOfSceneNames] = path
    const pathScene =
      scenes.find(({ name }) => {
        return firstSceneName.localeCompare(name) === 0
      }) ?? notFoundScene(move)
    const { actions } = pathScene
    return restOfSceneNames.length > 0 && actions.length > 0
      ? getSceneByPath(move)(actions, restOfSceneNames)
      : pathScene
  }

const getPath = ({ target }: Move): string[] =>
  target?.split('/').splice(1) ?? []

export const getTargetedScene: (
  move: Move
) => (scenes: BookScene[]) => BookScene = (move) => (scenes) => {
  return getSceneByPath(move)(scenes, getPath(move))
}

const updateState: (name: string) => (state: State) => State = (name) =>
  R.pipe(R.assoc('currentSceneName', name), incPlayedSceneCount(name))

const goto: (
  scenes: BookScene[],
  playedScenes: PlayedScene[]
) => [(move: Move) => boolean, (move: Move) => MovedScene] = (
  scenes,
  playedScenes
) => [
  R.propEq('type', 'anchor'),
  (move) => {
    const lastState = getLastPlayedSceneState(playedScenes)
    const scene = getTargetedScene(move)(scenes)
    const { name } = scene

    return {
      ...scene,
      state: updateState(name)(lastState)
    }
  }
]

export default goto
