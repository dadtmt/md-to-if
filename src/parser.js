import SimpleMarkdown from 'simple-markdown'

// String -> [String]
export const matchBraces = source => /^\{([\s\S]*?)\}/.exec(source)

// String -> [String]
export const matchBracketPipe = source => /^\[([\s\S]*?)\|/.exec(source)

// String -> [String]
export const matchPipeBracket = source => /^\|([\s\S]*?)\]/.exec(source)

const command = {
  order: 0,
  match: matchBraces,
  parse: (capture, parse, state) => ({ content: parse(capture[1], state) }),
}

const trueCaseContent = {
  order: 0,
  match: matchBracketPipe,
  parse: (capture, parse, state) => ({ content: parse(capture[1], state) }),
}

const falseCaseContent = {
  order: 0,
  match: matchPipeBracket,
  parse: (capture, parse, state) => ({ content: parse(capture[1], state) }),
}

// String -> [Content]
const parser = SimpleMarkdown.parserFor({
  ...SimpleMarkdown.defaultRules,
  command,
  trueCaseContent,
  falseCaseContent,
})

export default parser
