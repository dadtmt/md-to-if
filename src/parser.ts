import SimpleMarkdown, { SingleASTNode } from 'simple-markdown'

// String -> [String]
export const matchBraces = (source: string): RegExpExecArray | null =>
  /^\{([\s\S]*?)\}/.exec(source)

// String -> [String]
export const matchBracketPipe = (source: string): RegExpExecArray | null =>
  /^\[([\s\S]*?)\|/.exec(source)

// String -> [String]
export const matchPipeBracket = (source: string): RegExpExecArray | null =>
  /^\|([\s\S]*?)\]/.exec(source)

const parse: (capture: any, parse: any, state: any) => any = (
  capture,
  parse,
  state
) => ({ content: parse(capture[1], state) })

const command = {
  order: 0,
  match: matchBraces,
  parse
}

const trueCaseContent = {
  order: 0,
  match: matchBracketPipe,
  parse
}

const falseCaseContent = {
  order: 0,
  match: matchPipeBracket,
  parse
}

// String -> [Content]
const parser: (source: string) => SingleASTNode[] = SimpleMarkdown.parserFor({
  ...SimpleMarkdown.defaultRules,
  command,
  trueCaseContent,
  falseCaseContent
})

export default parser
