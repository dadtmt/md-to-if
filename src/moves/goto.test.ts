import parser from '../parser'
import { getTargetedScene } from './goto'
import adventure from '../adventure.md'
import book from '../book'
import { ActionScene } from '..'

const adventureBook = book(parser(adventure))

const moveToCantina = {
  type: 'anchor',
  target: '/cantina'
}
const moveToUnknown = {
  type: 'anchor',
  target: '/unknown'
}
const moveToCantinaDrink = {
  type: 'anchor',
  target: '/cantina/drink'
}

const cantinaScene = adventureBook[2]
const { actions } = cantinaScene as ActionScene
const [drinkActionScene] = actions

describe('getTargetedScene', () => {
  it('get the targeted scene', () => {
    expect(getTargetedScene(moveToCantina)(adventureBook)).toEqual(cantinaScene)
  })
  it('get the targeted action scene', () => {
    expect(getTargetedScene(moveToCantinaDrink)(adventureBook)).toEqual(
      drinkActionScene
    )
  })
  it('get the unknown scene', () => {
    expect(getTargetedScene(moveToUnknown)(adventureBook)).toMatchSnapshot()
  })
})
