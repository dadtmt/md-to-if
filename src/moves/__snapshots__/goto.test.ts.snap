// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getTargetedScene get the unknown scene 1`] = `
Object {
  "actions": Array [],
  "name": "unknown",
  "sceneContent": Array [
    Object {
      "content": Array [
        Object {
          "content": "unknown",
          "type": "text",
        },
      ],
      "level": 2,
      "type": "heading",
    },
    Object {
      "content": Array [
        Object {
          "content": "The scene with path: /unknown does not exist",
          "type": "text",
        },
      ],
      "type": "paragraph",
    },
  ],
}
`;
